Kernel/Memory/MemoryManager.cpp

```diff
@@ -190,9 +190,21 @@ UNMAP_AFTER_INIT void MemoryManager::register_reserved_ranges()
     m_reserved_memory_ranges.append(ContiguousReservedMemoryRange { range.start, m_physical_memory_ranges.last().start.get() + m_physical_memory_ranges.last().length - range.start.get() });
 }
 
-bool MemoryManager::is_allowed_to_read_physical_memory_for_userspace(PhysicalAddress start_address, size_t read_length) const
+bool MemoryManager::is_allowed_to_read_physical_memory_for_userspace(PhysicalMemoryExposeMethod expose_method, PhysicalAddress start_address, size_t read_length) const
 {
     VERIFY(!m_reserved_memory_ranges.is_empty());
+    // Note: This check happens to guard against possible memory leak.
+    // For example, if we try to mmap physical memory from 0x1000 to 0x2000 and you
+    // can actually mmap only from 0x1001, then we would fail as usual.
+    // However, in such case if we mmap from 0x1002, we are technically not violating
+    // any rules, besides the fact that we mapped an entire page with two bytes which we
+    // were not supposed to see. To prevent that, if we use mmap(2) syscall, we should
+    // always consider the start page to be aligned on PAGE_SIZE, or to be more precise
+    // is to be set to the page base of that start address.
+    if (expose_method == PhysicalMemoryExposeMethod::Mmap) {
+        VERIFY(start_address == start_address.page_base());
+    }
```

Comment:
useless curlies, but I'm not convinced that adding a whole enum argument that differentiates between the 2 callers just for this extra verify is the nicest way to do it, can't we just verify this at the start of the `mmap` method?

Reply:
We technically rely on this to ensure we don't `mmap(2)` on unaligned address:
https://github.com/SerenityOS/serenity/blob/168063819d009b34d43286c0367129592809a79d/Kernel/Syscalls/mmap.cpp#L158-L159
Still I see value in keeping this `VERIFY` in place, to ensure no matter what happens to `mmap(2)` we still `VERIFY` this. We could also verify this in the mmap method of the `MemoryDevice` (if that is what you actually intended to say).